### PR TITLE
Add the ability to build a relation in the inspector

### PR DESCRIPTION
**Pull request by hatched**

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5314/
Test FAILed.

**Comment by huwshimi** [comment]
A few QA things:
- when you select an application to relate to there is a console error: "Warning: Failed propType: Invalid prop `label` of type `object` supplied to `<<anonymous>>`, expected `string`."
- if you have multiple relation types (e.g. try and relate apache2 and wordpress) every checkbox you click selects the top checkbox
- if you relate mysql and wordpress and then try and build a relation the message is "No relatable endpoints for these applications." I'm not sure if that's quite correct, but maybe it's OK
- if you have several applications that can relate to each and have relations sometimes it seems to get into a state where it gives you the option to create a relation you've already created
- add two mysql apps, see two relation options, build a master relation, see that there is only the slave option, build it, go to build another relation and see that both options are now available

File: jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+          }
+        }
+      });
+    },
+
+    /**
+      Generate the list items from a set of services
+
+      @method generateItemList
+      @param {Object} services Relatable services.
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js]: No params for this function.

**Comment by anthonydillon** [comment]
Awesome stuff @hatched! All worked a dream but for one small QA issue:
- When going into a relate able service I get an error in the console `Warning: Failed propType: Invalid prop `label` of type `object` supplied to `Constructor`, expected `string`. Check the render method of `Constructor`.`

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5348/
Test FAILed.

**Comment by hatched** [comment]
Thanks for the reviews @anthonydillon and @huwshimi - I've now resolved all of the issues found in qa however interestingly the issues mentioned by @huwshimi also happen with the relations on the canvas so I'll need to do a follow-up which replaces the old logic with the new utils logic.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5349/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5370/
Test FAILed.

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2015 Canonical Ltd.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js]: 2016

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      changeState: React.PropTypes.func.isRequired,
+      createRelation: React.PropTypes.func.isRequired,
+      relationTypes: React.PropTypes.array.isRequired
+    },
+
+    /**
+      Generate the initial state.
+
+      @method getInitialState
+      @returns {String} The intial state.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js]: Object

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js [hunk 4]
```diff
@@ -0,0 +1,290 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2015 Canonical Ltd.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: 2016

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+describe('InspectorRelations', function() {
+
+  beforeAll(function(done) {
+    // By loading this file it adds the component to the juju components.
+    YUI().use('inspector-relations', function() { done(); });
+  });
+
+  it('can render the relations list', function() {
+    var changeState = sinon.stub();
+    var relations = [
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: I think we tend to do this kind of indentation like:
`var relations = [{`
and then between objects we do:
`}, {`

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+          role: 'primary'
+        },
+        far: {
+          name: 'django',
+          serviceName: 'django'
+        },
+        interface: 'postgresql',
+        scope: 'global'
+      },
+      {id: 'postgresql',
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: `}, {`?

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+          name: 'pgsql',
+          role: 'primary'
+        },
+        far: {
+          name: 'django',
+          serviceName: 'django'
+        },
+        interface: 'postgresql',
+        scope: 'global'}
+    ];
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: `}];`?

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    var buttons = [];
+    buttons.push({
+      title: 'Remove',
+      type: 'neutral',
+      action: instance._handleRemoveRelation,
+      disabled: true
+    });
+    var expected = (<div className="inspector-relations">
+      <ul className="inspector-relations__list">
+        <juju.components.InspectorRelationsItem
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: Oh, looks like these tests are not updated yet.

File: jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2015 Canonical Ltd.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js]: 2016

File: jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js (hunk 1 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js]: Also needs `@returns`

File: jujugui/static/gui/src/app/components/inspector-relate-to/test-inspector-relate-to.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2015 Canonical Ltd.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector-relate-to/test-inspector-relate-to.js]: 2016

File: jujugui/static/gui/src/app/views/utils.js [hunk 11]
```diff
@@ -1765,6 +1765,95 @@ YUI.add('juju-view-utils', function(Y) {
   };
 
   /**
+    Create a list of relations.
+
+    @method createRelation
+    @param {Object} env The current environment.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/views/utils.js]: Need db param.

File: jujugui/static/gui/src/app/views/utils.js [hunk 12]
```diff
@@ -1765,6 +1765,95 @@ YUI.add('juju-view-utils', function(Y) {
...
+      }
+    ]];
+    var relationId = 'pending-' + endpoints[0][0] + ':' + endpoints[0][1].name +
+                      endpoints[1][0] + ':' + endpoints[1][1].name;
+    db.relations.add({
+      relation_id: relationId,
+      'interface': endpoints[0][1].name,
+      endpoints: endpoints,
+      pending: true,
+      scope: 'global', // XXX check the charms to see if this is a subordinate
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/views/utils.js]: Does this need to be added to this branch or a card made?

**Comment by huwshimi** [comment]
This is looking great @hatched, QA looks OK and I started doing a bit of code review, but looks like the tests still need to be updated before a full review.

File: jujugui/static/gui/src/app/views/utils.js (hunk 12 above)
Comment (hatched) [inline jujugui/static/gui/src/app/views/utils.js]: card made, thanks for the reminder

File: jujugui/static/gui/src/app/views/utils.js (hunk 11 above)
Comment (hatched) [inline jujugui/static/gui/src/app/views/utils.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to/test-inspector-relate-to.js (hunk 10 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to/test-inspector-relate-to.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js (hunk 9 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js (hunk 8 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: Right, sorry, most of the tests were not rewritten yet.

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js (hunk 4 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/inspector-relate-to-type.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to/inspector-relate-to.js]: done

File: jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-relate-to-type/test-inspector-relate-to-type.js]: The two new component test suites appear to simply have been copied over from other components without changes so these need to be deleted and completely rewritten.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5377/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5378/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5379/
Test PASSed.

**Comment by huwshimi** [comment]
QA OK :+1: thanks for your persistence @hatched.

**Comment by anthonydillon** [comment]
Thank you for picking this one up and getting it done @hatched! Code looks good :+1:

**Comment by hatched** [comment]
Thanks for the reviews and qa's! 

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**